start_admin_sshd.sh

```diff
@@ -4,64 +4,122 @@
+      log "Failed to parse authorized_keys from ${user_data}"
+      return 1
+    fi
+
+    # Map the keys to avoid improper splitting
+    local mapped_keys
+    mapfile -t mapped_keys <<< "${raw_keys}"
+
+    # Verify the keys are valid
+    local key
+    local -a valid_keys
+    for key in "${mapped_keys[@]}"; do
+      if ! echo "${key}" | ssh-keygen -lf - &>/dev/null; then
+        log "Failed to validate ${key}"
+        continue
+      fi
+      valid_keys+=( "${key}" )
+    done
+
+    ( IFS=$'\n'; echo "${valid_keys[*]}" )
+}
+
+# Populate authorized_keys with all the authorized keys found in user-data
+if authorized_keys=$(parse_authorized_keys); then
+  ssh_authorized_keys="${user_ssh_dir}/authorized_keys"
+  install -D -m 0600 /dev/null "${ssh_authorized_keys}"
+  echo "${authorized_keys}" > "${ssh_authorized_keys}"
+  ((++available_auth_methods))
+fi
+
+parse_trusted_user_ca_keys() {
+    # Extract the keys from user-data json
+    local raw_keys
+    if ! raw_keys=$(jq -e -r '.["ssh"]["trusted_user_ca_keys"][]' "${user_data}" 2>/dev/null); then
+      log "Failed to parse trusted_user_ca_keys from ${user_data}"
+      return 1
+    fi
+
+    # Map the keys to avoid improper splitting
+    local mapped_keys
+    mapfile -t mapped_keys <<< "${raw_keys}"
+
+    # Verify the keys are valid
+    local key
+    local -a valid_keys
+    for key in "${mapped_keys[@]}"; do
+      if ! echo "${key}" | ssh-keygen -lf - &>/dev/null; then
+        log "Failed to validate ${key}"
+        continue
+      fi
+      valid_keys+=( "${key}" )
+    done
+
+    ( IFS=$'\n'; echo "${valid_keys[*]}" )
+}
+
+# Populate trusted_user_ca_keys with all the authorized keys found in user-data
+if trusted_user_ca_keys=$(parse_trusted_user_ca_keys); then
+  ssh_trusted_user_ca_keys="/etc/ssh/trusted_user_ca_keys.pub"
+  sshd_config="/etc/ssh/sshd_config"
+  install -D -m 0600 /dev/null "${ssh_trusted_user_ca_keys}"
+  echo "${trusted_user_ca_keys}" > "${ssh_trusted_user_ca_keys}"
+
+  cat >> "${sshd_config}" <<EOF
+
+# Configured by user data
+TrustedUserCAKeys ${ssh_trusted_user_ca_keys}
+EOF
+
+  ((++available_auth_methods))
 fi
 
-chown ec2-user -R "${ssh_config_dir}"
+chown ec2-user -R "${user_ssh_dir}"
+
+# If there were no successful auth methods, then users cannot authenticate
+if [[ "${available_auth_methods}" -eq 0 ]]; then
+  log "Failed to configure ssh authentication"
+  exit 1
+fi
 
 # Generate the server keys
 mkdir -p "${ssh_host_key_dir}"
 for key in rsa ecdsa ed25519; do
-    # If both of the keys exist, don't overwrite them
-    if [ -s "${ssh_host_key_dir}/ssh_host_${key}_key" ] && [ -s "${ssh_host_key_dir}/ssh_host_${key}_key.pub"  ]; then
-        echo "${key} key already exists, will use existing key." >&2
-        continue
-    fi
+  # If both of the keys exist, don't overwrite them
+  if [[ -s "${ssh_host_key_dir}/ssh_host_${key}_key" ]] &&
+  [[ -s "${ssh_host_key_dir}/ssh_host_${key}_key.pub" ]]; then
```

Comment:
We don't need to use `[[` here; seems like an unnecessary change. I prefer to use `[` unless we're taking advantage of specific features that only `[[` can provide.